kernel/arch/riscv/debug.c

```diff
@@ -13,27 +13,191 @@
 
 #include <elf.h>
 
+static bool fp_is_valid(ulong fp, ulong sp)
+{
+   ulong low = sp + 2 * sizeof(void *);
+   ulong high = (sp + (KERNEL_STACK_SIZE - 1)) & ~(KERNEL_STACK_SIZE - 1);
+
+   if (!fp || fp < low || fp > high || fp & 0x7)
+      return false;
+   else
+      return true;
+}
+
+static size_t
+stackwalk32(void **frames,
+            size_t count,
+            void *fp,
+            pdir_t *pdir)
+{
+   bool curr_pdir = false;
+   void *retAddr;
+   size_t i;
+   ulong old_fp;
+
+   if (!fp) {
+      fp = __builtin_frame_address(0);
+   }
+
+   if (!pdir) {
+      pdir = get_curr_pdir();
+      curr_pdir = true;
+   }
+
+   for (i = 0; i < count; i++) {
+
+      old_fp = (ulong)fp;
+
+      if ((ulong)fp < BASE_VA)
+         break;
+
+      if (curr_pdir) {
+
+         retAddr = *((void **)fp - 1);
+         fp = *((void **)fp - 2);
```

Comment:
Purely cosmetic: for symmetry, there should be an empty line before `} else {` since after `} else {` there is an empty line. Also, note that when the `if (...)` line is much shorter than the following statement, if you want you can also skip the empty line and pack things up, but in that case you should also pack the line after `} else {` for symmetry as well. Whatever you prefer.

The reason for the empty line is that a long `if` or `for` followed by a short statement like `void *p;` hides the first short statement visually:
```
if (something(x) && (something_else(x + y) || !special_condition)) {
   void *p;
   do_something_else();
   foo();
}
```

 I always found that hard to read and that's why I started adding those extra lines. But for short statements, that doesn't apply:
```
if (p) {
   call_long_function(a, b, c, d);
}
```
Because the following line stands out a lot.

I realize how complex that might look like to remember. It's simply that I prefer the code to super visually readable instead of being consistent within a simple set of rules. I guess that Tilck's coding style is almost consistent, but the "set of rules" is very non-trivial at least :-)

Reply:
Thank you for explaining it in detail!